tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15395,22 +15370,22 @@ bool Sema::DiagnoseHLSLDecl(Declarator &D, DeclContext *DC, Expr *BitWidth,
       virtual void diagnose(Sema &S, SourceLocation Loc, QualType T) {}
     } SD;
     RequireCompleteType(D.getLocStart(), qt, SD);
-    TypeDiagContext DiagContext = TypeDiagContext::CBuffersOrTBuffers;
-    if (isStatic)
-      DiagContext = TypeDiagContext::GlobalVariables;
-    if (DiagnoseTypeElements(*this, D.getLocStart(), qt, DiagContext))
-      result = false;
-  }
 
-  // Disallow long vecs from $Global cbuffers.
-  if (isGlobal && !isStatic && !isGroupShared && !IS_BASIC_OBJECT(basicKind)) {
-    const unsigned CbuffersOrTbuffersIdx = 4;
-    if (ContainsLongVector(qt)) {
-      Diag(D.getLocStart(), diag::err_hlsl_unsupported_long_vector)
-          << CbuffersOrTbuffersIdx;
+    // Disallow objects in the global context
+    TypeDiagContext ObjDiagContext = TypeDiagContext::CBuffersOrTBuffers;
+    if (isGroupShared)
+      ObjDiagContext = TypeDiagContext::GroupShared;
+    else if (isStatic)
+      ObjDiagContext = TypeDiagContext::GlobalVariables;
 
+    TypeDiagContext LongVecDiagContext = TypeDiagContext::Valid;
+
+    // Disallow long vecs from $Global cbuffers.
+    if (!isStatic && !isGroupShared && !IS_BASIC_OBJECT(basicKind))
+      LongVecDiagContext = TypeDiagContext::CBuffersOrTBuffers;
```

Comment:
I was imagining `else if (!IS_BASIC_OBJECT(basicKind))` as the final else if after the `else if (isStatic)` on line 15378 which would result in something more concise, but I won't insist. If you make another change to the PR, you might consider that consolidation.